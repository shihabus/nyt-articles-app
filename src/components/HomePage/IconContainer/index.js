import React from "react";
import { View, StyleSheet } from "react-native";
import Icon from "react-native-vector-icons/FontAwesome";

const styles = StyleSheet.create({
  container: {
    width: "10%",
    flexDirection: "row",
    justifyContent: "center",
    alignContent: "center"
  },
  icon: { alignSelf: "center" }
});

const IconContainer = () => {
  return (
    <View style={styles.container}>
      <Icon name="chevron-right" size={15} color="gray" style={styles.icon} />
    </View>
  );
};

export default IconContainer;
