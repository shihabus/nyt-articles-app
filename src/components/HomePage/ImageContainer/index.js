import React from 'react'
import { View, StyleSheet,Image } from 'react-native'
import PropTypes from 'prop-types';

const styles=StyleSheet.create({
    imageContainer: {
        flexDirection: "row",
        justifyContent: "center",
        width: "20%"
      },
      imageStyle: {
        backgroundColor: "#CED0CE",
        height: 50,
        width: 50,
        alignSelf: "center",
        borderRadius: 25
      },
})
const ImageContainer = ({ urlToImage }) => (
    <View style={styles.imageContainer}>
      <Image
        style={styles.imageStyle}
        resizeMode="contain"
        source={{ uri: urlToImage }}
      />
    </View>
  );

ImageContainer.propTypes={
    urlToImage:PropTypes.string
}

export default ImageContainer
