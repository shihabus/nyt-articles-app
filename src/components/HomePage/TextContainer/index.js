/* eslint-disable camelcase */
import React from "react";
import { View, StyleSheet } from "react-native";
import PropTypes from "prop-types";
import Author from './Author'
import Footer from './Footer'
import Title from './Title'

const styles=StyleSheet.create({
  container: {
    flexDirection: "column",
    justifyContent: "center",
    width: "70%"
  },
  subContainer: {
    paddingVertical: 5
  },
})
const TextContainer = ({ params }) => {
  const { title, byline, published_date, source } = params;
  return (
    <View style={styles.container}>
      <View style={styles.subContainer}>
        <Title title={title} />
      </View>
      <View style={styles.subContainer}>
        <Author byline={byline} />
        <Footer published_date={published_date} source={source} />
      </View>
    </View>
  );
};

TextContainer.propTypes = {
  params: PropTypes.shape({
    title: PropTypes.string,
    byline: PropTypes.string,
    published_date: PropTypes.string,
    source: PropTypes.string
  })
};

export default TextContainer;
