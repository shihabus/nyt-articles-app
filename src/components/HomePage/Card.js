/* eslint-disable camelcase */
import React from "react";
import PropTypes from "prop-types";
import { TouchableOpacity, StyleSheet } from "react-native";
import ImageContainer from "./ImageContainer";
import TextContainer from "./TextContainer";
import IconContainer from "./IconContainer";

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    justifyContent: "space-between",
    paddingVertical: 20,
    backgroundColor: "#fff",
    alignContent: "center"
  }
});

export const ImageExtractor=(media,type)=>(
  media[0]["media-metadata"]
  .filter(x => x.format === type)[0]
  .url.toString()
)

const Card = ({ item,handleSelection }) => {

  const { byline, source, published_date, title, media,abstract,url } = item;

  const urlToImage = ImageExtractor(media,"Large Thumbnail")

  const urlToLargeImage = ImageExtractor(media,"mediumThreeByTwo440")

  const params = { byline, source, published_date, title,abstract,url,imageURL:urlToLargeImage};

  const handler=(selection)=>{
    handleSelection(selection)
  }

  return (
    <TouchableOpacity
      style={styles.container}
      onPress={()=>handler(params)}
    >
      <ImageContainer urlToImage={urlToImage} testID='ImageContainer'/>
      <TextContainer params={params} testID='TextContainer'/>
      <IconContainer testID='IconContainer'/>
    </TouchableOpacity>
  );
};

Card.propTypes = {
  item: PropTypes.shape({
    byline: PropTypes.string,
    source: PropTypes.string,
    published_date: PropTypes.string,
    title: PropTypes.string,
    media: PropTypes.arrayOf(PropTypes.object),
    id: PropTypes.number,
    abstract: PropTypes.string,
    url: PropTypes.string
  }),
  handleSelection: PropTypes.func
};

export default Card;
