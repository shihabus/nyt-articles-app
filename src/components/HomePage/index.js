/* eslint-disable react/prop-types */
import React from "react";
import { View, FlatList, StyleSheet } from "react-native";
import PropTypes from "prop-types";
import Card from "./Card";

const styles = StyleSheet.create({
  itemLineSeparator: {
    height: 0.5,
    width: "100%",
    backgroundColor: "#CED0CE",
    marginLeft: 10
  }
});


export const List = ({ article, handleSelection }) => {
  const renderSeparator = () => <View style={styles.itemLineSeparator} />;

  const renderArticle = ({ item }) => {
    const { id } = item;
    return <Card handleSelection={handleSelection} item={item} key={id} />;
  };

  return (
    <View>
      <FlatList
        data={article}
        contentContainerStyle={{ paddingBottom: 5 }}
        ItemSeparatorComponent={renderSeparator}
        renderItem={x => renderArticle(x)}
      />
    </View>
  );
};

List.propTypes = {
  article: PropTypes.arrayOf(PropTypes.object),
  handleSelection: PropTypes.func,
};

