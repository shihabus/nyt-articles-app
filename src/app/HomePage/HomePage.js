/* eslint-disable react/sort-comp */
/* eslint-disable no-useless-constructor */
import React, { Component } from "react";
import Icon from "react-native-vector-icons/FontAwesome";
import { View, StyleSheet, SafeAreaView } from "react-native";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { fetchNews, selectedArticle } from "./store";
import { Error, Loader } from "../../components/Common";
import { List } from "../../components/HomePage";
import { COLORS, STRINGS } from "../../constants";

const styles = StyleSheet.create({
  container: {
    flex: 1
  }
});

class HomePage extends Component {

  constructor(props) {
    super(props)
}

  static navigationOptions = {
    title: STRINGS.homePageTitle,
    headerRight: (
      <View style={{ paddingHorizontal: 20 }}>
        <Icon name="search" size={20} color={COLORS.headerTextColor} />
      </View>
    ),
    headerLeft: (
      <View style={{ paddingHorizontal: 20 }}>
        <Icon name="bars" size={20} color={COLORS.headerTextColor} />
      </View>
    ),
    headerStyle: {
      backgroundColor: COLORS.headerBar
    },
    headerTintColor: COLORS.headerTextColor,
    headerTitleStyle: {
      fontWeight: "normal",
      fontSize: 18
    }
  };

  componentDidMount() {
    const { fetchNews } = this.props;
    fetchNews();
  }

  selectionHandler = item => {
    const { selectedArticle, navigation } = this.props;
    selectedArticle(item);
    navigation.navigate("Details");
  };

  renderSuspense = () => {
    const { error, loading, articleArray } = this.props;
    if (error) {
      return <Error testID='Error'/>;
    }
    if (loading) {
      return <Loader testID='Loader'/>;
    }
    return (
      <View style={{ flex: 1 }} testID='List'>
        <List
          article={articleArray}
          handleSelection={this.selectionHandler}
        />
      </View>
    );
  };

  render() {
    return (
      <SafeAreaView style={styles.container}>
        {this.renderSuspense()}
      </SafeAreaView>
    );
  }
}

HomePage.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  navigation: PropTypes.object,
  fetchNews: PropTypes.func,
  articleArray: PropTypes.arrayOf(PropTypes.object),
  error: PropTypes.bool,
  loading: PropTypes.bool,
  selectedArticle: PropTypes.func
};

const mapStateToProps = ({ articles }) => {
  const { loading, articleArray, error, success } = articles;
  return { loading, articleArray, error, success };
};

export default connect(
  mapStateToProps,
  { fetchNews, selectedArticle }
)(HomePage);


