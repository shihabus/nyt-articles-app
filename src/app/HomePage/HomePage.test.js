/* eslint-disable no-unused-vars */
/* eslint-disable import/named */
/* eslint-disable no-undef */
/**
 * @jest-environment jsdom
 */

import React from "react";
import { createSerializer } from "enzyme-to-json";
import { connect } from "react-redux";
import HomePage, { mapStateToProps } from "./HomePage";
import { findComponent, testStore, checkProps } from "../../../testUtils";

expect.addSnapshotSerializer(createSerializer({ mode: "deep" }));

const setUp = (initialStore = {}) => {
  const store = testStore(initialStore);
  const wrapper = shallow(<HomePage store={store} />)
    .childAt(0)
    .dive();
  return wrapper;
};

describe("Home Page", () => {
  let wrapper;
  beforeEach(() => {
    const initialStore = {
      selection: {
        articles: {
          loading: true,
          articleArray: [],
          error: false,
          success: false
        }
      }
    };
    wrapper = setUp(initialStore);
  });

  it("Snapshot", () => {
    expect(wrapper).toMatchSnapshot("Home page");
  });

  it("Checked props", () => {
    const expectedProps = {
      fetchNews: () => {},
      articleArray: [{ id: 1 }, { id: 2 }],
      error: true,
      loading: 1,
      selectedArticle: () => {}
    };
    const propsErr = checkProps(HomePage, expectedProps);
    expect(propsErr).toBeUndefined();
  });

  it("should render List component on success", () => {
    expect(findComponent(wrapper, "List").length).toEqual(1);
  });

  //   it("should render Error component", () => {
  //     // expect(findComponent(wrapper, "Error").length).toEqual(1);
  //     expect(1).toEqual(1)

  //   });

  //   it("should render Loader component", () => {
  //     // expect(findComponent(wrapper, "Loader").length).toEqual(1);
  //     expect(1).toEqual(1)
  //   });
});
