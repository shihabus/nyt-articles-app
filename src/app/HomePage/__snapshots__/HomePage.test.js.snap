// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home Page Snapshot: Home page 1`] = `
<SafeAreaView
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
    testID="List"
  >
    <List
      handleSelection={[Function]}
    />
  </View>
</SafeAreaView>
`;
